<!DOCTYPE html>
<html lang="en"><head><meta charset="UTF-8"/><meta name="viewport" content="width=device-width, initial-scale=1.0"/><title>FASTA · FASTX.jl</title><meta name="title" content="FASTA · FASTX.jl"/><meta property="og:title" content="FASTA · FASTX.jl"/><meta property="twitter:title" content="FASTA · FASTX.jl"/><meta name="description" content="Documentation for FASTX.jl."/><meta property="og:description" content="Documentation for FASTX.jl."/><meta property="twitter:description" content="Documentation for FASTX.jl."/><script data-outdated-warner src="../assets/warner.js"></script><link href="https://cdnjs.cloudflare.com/ajax/libs/lato-font/3.0.0/css/lato-font.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/juliamono/0.050/juliamono.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/fontawesome.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/solid.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/6.4.2/css/brands.min.css" rel="stylesheet" type="text/css"/><link href="https://cdnjs.cloudflare.com/ajax/libs/KaTeX/0.16.8/katex.min.css" rel="stylesheet" type="text/css"/><script>documenterBaseURL=".."</script><script src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.6/require.min.js" data-main="../assets/documenter.js"></script><script src="../search_index.js"></script><script src="../siteinfo.js"></script><script src="../../versions.js"></script><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-dark.css" data-theme-name="documenter-dark" data-theme-primary-dark/><link class="docs-theme-link" rel="stylesheet" type="text/css" href="../assets/themes/documenter-light.css" data-theme-name="documenter-light" data-theme-primary/><script src="../assets/themeswap.js"></script></head><body><div id="documenter"><nav class="docs-sidebar"><a class="docs-logo" href="../"><img src="../assets/logo.png" alt="FASTX.jl logo"/></a><div class="docs-package-name"><span class="docs-autofit"><a href="../">FASTX.jl</a></span></div><button class="docs-search-query input is-rounded is-small is-clickable my-2 mx-auto py-1 px-2" id="documenter-search-query">Search docs (Ctrl + /)</button><ul class="docs-menu"><li><span class="tocitem">Overview</span><ul><li><a class="tocitem" href="../">Quickstart</a></li><li><a class="tocitem" href="../records/">Records</a></li><li><a class="tocitem" href="../files/">File I/O</a></li></ul></li><li class="is-active"><a class="tocitem" href>FASTA</a><ul class="internal"><li><a class="tocitem" href="#The-FASTARecord"><span>The <code>FASTARecord</code></span></a></li><li><a class="tocitem" href="#FASTAReader-and-FASTAWriter"><span><code>FASTAReader</code> and <code>FASTAWriter</code></span></a></li></ul></li><li><a class="tocitem" href="../fastq/">FASTQ</a></li><li><a class="tocitem" href="../fai/">FAI</a></li></ul><div class="docs-version-selector field has-addons"><div class="control"><span class="docs-label button is-static is-size-7">Version</span></div><div class="docs-selector control is-expanded"><div class="select is-fullwidth is-size-7"><select id="documenter-version-selector"></select></div></div></div></nav><div class="docs-main"><header class="docs-navbar"><a class="docs-sidebar-button docs-navbar-link fa-solid fa-bars is-hidden-desktop" id="documenter-sidebar-button" href="#"></a><nav class="breadcrumb"><ul class="is-hidden-mobile"><li class="is-active"><a href>FASTA</a></li></ul><ul class="is-hidden-tablet"><li class="is-active"><a href>FASTA</a></li></ul></nav><div class="docs-right"><a class="docs-navbar-link" href="https://github.com/BioJulia/FASTX.jl" title="View the repository on GitHub"><span class="docs-icon fa-brands"></span><span class="docs-label is-hidden-touch">GitHub</span></a><a class="docs-navbar-link" href="https://github.com/BioJulia/FASTX.jl/blob/master/docs/src/fasta.md" title="Edit source on GitHub"><span class="docs-icon fa-solid"></span></a><a class="docs-settings-button docs-navbar-link fa-solid fa-gear" id="documenter-settings-button" href="#" title="Settings"></a><a class="docs-article-toggle-button fa-solid fa-chevron-up" id="documenter-article-toggle-button" href="javascript:;" title="Collapse all docstrings"></a></div></header><article class="content" id="documenter-page"><h1 id="FASTA-formatted-files"><a class="docs-heading-anchor" href="#FASTA-formatted-files">FASTA formatted files</a><a id="FASTA-formatted-files-1"></a><a class="docs-heading-anchor-permalink" href="#FASTA-formatted-files" title="Permalink"></a></h1><p><strong>NB: First read the overview in the sidebar</strong></p><p>FASTA is a text-based file format for representing biological sequences. A FASTA file stores a list of sequence records with name, description, and sequence.</p><p>The template of a sequence record is:</p><pre><code class="nohighlight hljs">&gt;{description}
{sequence}</code></pre><p>Where the &quot;identifier&quot; is the first part of the description up to the first whitespace (or the entire description if there is no whitespace)</p><p>Here is an example of a chromosomal sequence:</p><pre><code class="nohighlight hljs">&gt;chrI chromosome 1
CCACACCACACCCACACACCCACACACCACACCACACACCACACCACACC
CACACACACACATCCTAACACTACCCTAACACAGCCCTAATCTA</code></pre><p>Here:</p><ul><li>The <code>identifier</code> is <code>&quot;chrI&quot;</code></li><li>The <code>description</code> is <code>&quot;chrI chromosome 1&quot;</code>, containing the identifier</li><li>The sequence is the DNA sequence <code>&quot;CCACA...&quot;</code></li></ul><h2 id="The-FASTARecord"><a class="docs-heading-anchor" href="#The-FASTARecord">The <code>FASTARecord</code></a><a id="The-FASTARecord-1"></a><a class="docs-heading-anchor-permalink" href="#The-FASTARecord" title="Permalink"></a></h2><p>FASTA records are, by design, very lax in what they can contain. They can contain almost arbitrary byte sequences, including invalid unicode, and trailing whitespace on their sequence lines, which will be interpreted as part of the sequence. If you want to have more certainty about the format, you can either check the content of the sequences with a regex, or (preferably), convert them to the desired <code>BioSequence</code> type.</p><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="FASTX.FASTA.Record" href="#FASTX.FASTA.Record"><code>FASTX.FASTA.Record</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">FASTA.Record</code></pre><p>Mutable struct representing a FASTA record as parsed from a FASTA file. The content of the record can be queried with the following functions: <code>identifier</code>, <code>description</code>, <code>sequence</code>.</p><p>FASTA records are un-typed, i.e. they are agnostic to what kind of data they contain.</p><p>See also: <a href="#FASTX.FASTA.Reader"><code>FASTA.Reader</code></a>, <a href="#FASTX.FASTA.Writer"><code>FASTA.Writer</code></a></p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; rec = parse(FASTARecord, &quot;&gt;some header\nTAqA\nCC&quot;);

julia&gt; identifier(rec)
&quot;some&quot;

julia&gt; description(rec)
&quot;some header&quot;

julia&gt; sequence(rec)
&quot;TAqACC&quot;

julia&gt; typeof(description(rec)) == typeof(sequence(rec)) &lt;: AbstractString
true</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/ab8159eb723c5cd44b3f833d9cd50b4079a27ffe/src/fasta/record.jl#L4-L31">source</a></section></article><h2 id="FASTAReader-and-FASTAWriter"><a class="docs-heading-anchor" href="#FASTAReader-and-FASTAWriter"><code>FASTAReader</code> and <code>FASTAWriter</code></a><a id="FASTAReader-and-FASTAWriter-1"></a><a class="docs-heading-anchor-permalink" href="#FASTAReader-and-FASTAWriter" title="Permalink"></a></h2><p><code>FASTAWriter</code> can optionally be passed the keyword <code>width</code> to control the line width. If this is zero or negative, it will write all record sequences on a single line. Else, it will wrap lines to the given maximal width.</p><h3 id="Reference:"><a class="docs-heading-anchor" href="#Reference:">Reference:</a><a id="Reference:-1"></a><a class="docs-heading-anchor-permalink" href="#Reference:" title="Permalink"></a></h3><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="FASTX.FASTA" href="#FASTX.FASTA"><code>FASTX.FASTA</code></a> — <span class="docstring-category">Module</span></header><section><div><pre><code class="language-julia hljs">FASTA</code></pre><p>Module under FASTX with code related to FASTA files.</p></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/ab8159eb723c5cd44b3f833d9cd50b4079a27ffe/src/fasta/fasta.jl#L4-L8">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="FASTX.FASTA.Reader" href="#FASTX.FASTA.Reader"><code>FASTX.FASTA.Reader</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">FASTA.Reader(input::IO; index=nothing, copy::Bool=true)</code></pre><p>Create a buffered data reader of the FASTA file format. The reader is a <code>BioGenerics.IO.AbstractReader</code>, a stateful iterator of <code>FASTA.Record</code>. Readers take ownership of the underlying IO. Mutating or closing the underlying IO not using the reader is undefined behaviour. Closing the Reader also closes the underlying IO.</p><p>See more examples in the FASTX documentation.</p><p>See also: <a href="#FASTX.FASTA.Record"><code>FASTA.Record</code></a>, <a href="#FASTX.FASTA.Writer"><code>FASTA.Writer</code></a></p><p><strong>Arguments</strong></p><ul><li><code>input</code>: data source</li><li><code>index</code>: Optional random access index (currently <em>fai</em> is supported). <code>index</code> can be <code>nothing</code>, a <code>FASTA.Index</code>, or an <code>IO</code> in which case an index will be parsed from the IO, or <code>AbstractString</code>, in which case it will be treated as a path to a fai file.</li><li><code>copy::Bool</code>: iterating returns fresh copies instead of the same Record. Set to <code>false</code> for improved performance, but be wary that iterating mutates records.</li></ul><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; rdr = FASTAReader(IOBuffer(&quot;&gt;header\nTAG\n&gt;another\nAGA&quot;));

julia&gt; records = collect(rdr); close(rdr);

julia&gt; foreach(println, map(identifier, records))
header
another

julia&gt; foreach(println, map(sequence, records))
TAG
AGA</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/ab8159eb723c5cd44b3f833d9cd50b4079a27ffe/src/fasta/reader.jl#L4-L40">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="FASTX.FASTA.Writer" href="#FASTX.FASTA.Writer"><code>FASTX.FASTA.Writer</code></a> — <span class="docstring-category">Type</span></header><section><div><pre><code class="language-julia hljs">FASTA.Writer(output::IO; width=70)</code></pre><p>Create a data writer of the FASTA file format. The writer is a <code>BioGenerics.IO.AbstractWriter</code>. Writers take ownership of the underlying IO. Mutating or closing the underlying IO not using the writer is undefined behaviour. Closing the writer also closes the underlying IO.</p><p>See more examples in the FASTX documentation.</p><p>See also: <a href="#FASTX.FASTA.Record"><code>FASTA.Record</code></a>, <a href="#FASTX.FASTA.Reader"><code>FASTA.Reader</code></a></p><p><strong>Arguments</strong></p><ul><li><code>output</code>: Data sink to write to</li><li><code>width</code>: Wrapping width of sequence characters. If &lt; 1, no wrapping.</li></ul><p><strong>Examples</strong></p><pre><code class="nohighlight hljs">julia&gt; FASTA.Writer(open(&quot;some_file.fna&quot;, &quot;w&quot;)) do writer
    write(writer, record) # a FASTA.Record
end</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/ab8159eb723c5cd44b3f833d9cd50b4079a27ffe/src/fasta/writer.jl#L4-L27">source</a></section></article><article class="docstring"><header><a class="docstring-article-toggle-button fa-solid fa-chevron-down" href="javascript:;" title="Collapse docstring"></a><a class="docstring-binding" id="FASTX.FASTA.validate_fasta" href="#FASTX.FASTA.validate_fasta"><code>FASTX.FASTA.validate_fasta</code></a> — <span class="docstring-category">Function</span></header><section><div><pre><code class="language-julia hljs">validate_fasta(io::IO) &gt;: Nothing</code></pre><p>Check if <code>io</code> is a valid FASTA file. Return <code>nothing</code> if it is, and an instance of another type if not.</p><p><strong>Examples</strong></p><pre><code class="language-julia-repl hljs">julia&gt; validate_fasta(IOBuffer(&quot;&gt;a bc\nTAG\nTA&quot;)) === nothing
true

julia&gt; validate_fasta(IOBuffer(&quot;&gt;a bc\nT&gt;G\nTA&quot;)) === nothing
false</code></pre></div><a class="docs-sourcelink" target="_blank" href="https://github.com/BioJulia/FASTX.jl/blob/ab8159eb723c5cd44b3f833d9cd50b4079a27ffe/src/fasta/readrecord.jl#L122-L136">source</a></section></article></article><nav class="docs-footer"><a class="docs-footer-prevpage" href="../files/">« File I/O</a><a class="docs-footer-nextpage" href="../fastq/">FASTQ »</a><div class="flexbox-break"></div><p class="footer-message">Powered by <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> and the <a href="https://julialang.org/">Julia Programming Language</a>.</p></nav></div><div class="modal" id="documenter-settings"><div class="modal-background"></div><div class="modal-card"><header class="modal-card-head"><p class="modal-card-title">Settings</p><button class="delete"></button></header><section class="modal-card-body"><p><label class="label">Theme</label><div class="select"><select id="documenter-themepicker"><option value="auto">Automatic (OS)</option><option value="documenter-light">documenter-light</option><option value="documenter-dark">documenter-dark</option></select></div></p><hr/><p>This document was generated with <a href="https://github.com/JuliaDocs/Documenter.jl">Documenter.jl</a> version 1.4.1 on <span class="colophon-date" title="Monday 24 June 2024 12:29">Monday 24 June 2024</span>. Using Julia version 1.10.4.</p></section><footer class="modal-card-foot"></footer></div></div></div></body></html>
